File: csharp/test/Ice/interceptor/InterceptorI.cs
// Copyright (c) ZeroC, Inc.

namespace Ice
{
    namespace interceptor
    {
        internal class InterceptorI : Ice.DispatchInterceptor
        {
            internal InterceptorI(Ice.Object servant)
            {
                servant_ = servant;
            }

            protected static void
            test(bool b)
            {
                if (!b)
                {
                    throw new System.Exception();
                }
            }

            public override Task<Ice.OutputStream>
            dispatch(Ice.Request request)
            {
                Ice.Current current = request.getCurrent();

                string context;
                if (current.ctx.TryGetValue("raiseBeforeDispatch", out context))
                {
                    if (context == "user")
                    {
                        throw new Test.InvalidInputException();
                    }
                    else if (context == "notExist")
                    {
                        throw new Ice.ObjectNotExistException();
                    }
                }

                lastOperation_ = current.operation;

                if (lastOperation_ == "addWithRetry" || lastOperation_ == "amdAddWithRetry")
                {
                    for (int i = 0; i < 10; ++i)
                    {
                        try
                        {
                            var t = servant_.ice_dispatch(request);
                            if (t != null && t.IsFaulted)
                            {
                                throw t.Exception.InnerException;
                            }
                            else
                            {
                                test(false);
                            }
                        }
                        catch (MyRetryException)
                        {
                            //
                            // Expected, retry
                            //
                        }
                    }

                    current.ctx["retry"] = "no";
                }
                else if (current.ctx.TryGetValue("retry", out context) && context == "yes")
                {
                    //
                    // Retry the dispatch to ensure that abandoning the result of the dispatch
                    // works fine and is thread-safe
                    //
                    servant_.ice_dispatch(request);
                    servant_.ice_dispatch(request);
                }

                var task = servant_.ice_dispatch(request);
                lastStatus_ = task != null;

                if (current.ctx.TryGetValue("raiseAfterDispatch", out context))
                {
                    if (context == "user")
                    {
                        throw new Test.InvalidInputException();
                    }
                    else if (context == "notExist")
                    {
                        throw new Ice.ObjectNotExistException();
                    }
                }

                return task;
            }

            internal bool
            getLastStatus()
            {
                return lastStatus_;
            }

            internal String
            getLastOperation()
            {
                return lastOperation_;
            }

            internal virtual void
            clear()
            {
                lastOperation_ = null;
                lastStatus_ = false;
            }

            protected readonly Ice.Object servant_;
            protected string lastOperation_;
            protected bool lastStatus_ = false;
        }
    }
}

File: cpp/test/Ice/interceptor/InterceptorI.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <InterceptorI.h>
#include <MyObjectI.h>
#include <TestHelper.h>

using namespace std;
using namespace Ice;

InterceptorI::InterceptorI(const ObjectPtr& servant) :
    _servant(servant),
    _lastStatus(false)
{
}

bool
InterceptorI::dispatch(Request& request)
{
    Current& current = const_cast<Current&>(request.getCurrent());

    Context::const_iterator p = current.ctx.find("raiseBeforeDispatch");
    if(p != current.ctx.end())
    {
        if(p->second == "user")
        {
            throw Test::InvalidInputException();
        }
        else if(p->second == "notExist")
        {
            throw ObjectNotExistException(__FILE__, __LINE__);
        }
        else if(p->second == "system")
        {
            throw MySystemException(__FILE__, __LINE__);
        }
    }

    _lastOperation = current.operation;

    if(_lastOperation == "addWithRetry")
    {
        for(int i = 0; i < 10; ++i)
        {
            try
            {
                _servant->ice_dispatch(request);
                test(false);
            }
            catch(const MyRetryException&)
            {
                //
                // Expected, retry
                //
            }
        }

        current.ctx["retry"] = "no";
    }

    _lastStatus = _servant->ice_dispatch(request);

    p = current.ctx.find("raiseAfterDispatch");
    if(p != current.ctx.end())
    {
        if(p->second == "user")
        {
            throw Test::InvalidInputException();
        }
        else if(p->second == "notExist")
        {
            throw ObjectNotExistException(__FILE__, __LINE__);
        }
        else if(p->second == "system")
        {
            throw MySystemException(__FILE__, __LINE__);
        }
    }

    return _lastStatus;
}

bool
InterceptorI::getLastStatus() const
{
    return _lastStatus;
}

const string&
InterceptorI::getLastOperation() const
{
    return _lastOperation;
}

void
InterceptorI::clear()
{
    _lastStatus = false;
    _lastOperation = "";
}
